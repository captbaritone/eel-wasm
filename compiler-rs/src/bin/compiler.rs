use eel_wasm::compile;
use std::fs;
use std::process;

use std::path::PathBuf;
use structopt::StructOpt;

#[derive(Debug, StructOpt)]
#[structopt(name = "eel-wasm", about = "Compile Eel code to WebAssembly.")]
struct Opt {
    /// Input file
    #[structopt(parse(from_os_str))]
    input: PathBuf,

    /// Output file, stdout if not present
    #[structopt(parse(from_os_str))]
    output: PathBuf,
}

fn main() {
    let opt = Opt::from_args();
    let filename = opt.input;
    let source = fs::read_to_string(filename).unwrap_or_else(|err| {
        eprintln!("Error reading file: {}", err);
        process::exit(1);
    });

    let result = compile(&source, vec![]).unwrap_or_else(|err| {
        eprintln!("{}", err);
        process::exit(1);
    });

    fs::write(opt.output, result).unwrap_or_else(|err| {
        eprintln!("Error writing output: {}", err);
        process::exit(1);
    });

    println!("Done.");
}
