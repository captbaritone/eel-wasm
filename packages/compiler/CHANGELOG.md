# Changelog

## 0.0.15

* Fix: Avoid crashing when using mod (`%`) with value between 0 and 1. [#54](https://github.com/captbaritone/eel-wasm/issues/54)

## 0.0.14

Initial shipped release. Shipped in Butterchurn and on Webamp.org.