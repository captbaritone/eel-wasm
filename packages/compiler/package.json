{
  "name": "eel-wasm",
  "version": "0.0.15",
  "main": "dist/src/index.js",
  "license": "MIT",
  "homepage": "https://github.com/captbaritone/eel-wasm",
  "author": "Jordan Eldredge <jordan@jordaneldredge.com>",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/captbaritone/eel-wasm.git"
  },
  "prettier": {
    "trailingComma": "es5"
  },
  "files": [
    "dist/**/*"
  ],
  "scripts": {
    "test": "jest",
    "ci": "yarn test && yarn build",
    "benchmark": "node tools/benchmark/performance.js perf",
    "consistency": "node tools/benchmark/performance.js consistency",
    "test-parse": "node tools/parseMilk.js --stats --summary --dir=./fixtures/mega",
    "test-compile": "node tools/parseMilk.js --stats --summary --compile --dir=./fixtures/mega",
    "build-parser": "node ./tools/buildParser.js > ./build/parser.js",
    "build": "yarn build-parser && tsc && parcel build src/index.ts -o dist/index.bundle.js --experimental-scope-hoisting",
    "clean": "rm -rf build/*",
    "prepublishOnly": "yarn test && yarn clean && yarn build"
  },
  "devDependencies": {
    "@babel/core": "^7.9.0",
    "@babel/preset-env": "^7.9.5",
    "@babel/preset-typescript": "^7.9.0",
    "@types/jest": "^25.2.1",
    "babel-jest": "^25.3.0",
    "jest": "^24.9.0",
    "jison": "^0.4.18",
    "milkdrop-eel-parser": "^0.0.4",
    "milkdrop-preset-utils": "^0.1.0",
    "parcel": "^1.12.4",
    "prettier": "^1.19.1",
    "print-diff": "^1.0.0",
    "typescript": "^3.8.3",
    "wabt": "^1.0.12",
    "yargs": "^15.1.0"
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "ts",
      "eel"
    ],
    "testPathIgnorePatterns": [
      "/node_modules/",
      "dist"
    ]
  }
}
