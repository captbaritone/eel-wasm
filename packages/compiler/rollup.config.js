import typescript from "@rollup/plugin-typescript";
import terser from "@rollup/plugin-terser";

export default {
  input: "src/index.ts",
  output: {
    dir: "dist",
    format: "es",
    sourcemap: true,
  },
  plugins: [typescript(), terser()],
};
