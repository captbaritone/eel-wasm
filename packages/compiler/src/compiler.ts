import { parse } from "./parser";
import { emit } from "./emitter";
import {
  encodeString,
  encodeNestedVector,
  encodeSection,
  op,
  unsignedLEB128,
  VAL_TYPE,
  FUNCTION_TYPE,
  GLOBAL_TYPE,
  MUTABILITY,
  TYPE_IDX,
  EXPORT_TYPE,
  SECTION,
  MAGIC,
  WASM_VERSION,
  encodeFlatVector,
} from "./encoding";
import shims from "./shims";
import * as Utils from "./utils";
import { localFuncMap } from "./wasmFunctions";
import { CompilerContext, TypedFunction, EelVersion } from "./types";
import { WASM_MEMORY_SIZE } from "./constants";

type CompilerOptions = {
  pools: {
    [name: string]: Set<string>;
  };
  functions: {
    [name: string]: {
      pool: string;
      code: string;
    };
  };
  eelVersion?: EelVersion;
  preParsed?: boolean;
};

export function compileModule({
  pools,
  functions: funcs,
  eelVersion = 2,
  preParsed = false,
}: CompilerOptions) {
  if (Object.keys(pools).includes("shims")) {
    throw new Error(
      'You may not name a pool "shims". "shims" is reserved for injected JavaScript functions.'
    );
  }

  // Collect all the globals that we expect to get as imports.
  const importedVars: [string, string][] = [];
  Object.entries(pools).forEach(([poolName, pool]) => {
    pool.forEach(variableName => {
      importedVars.push([poolName, variableName]);
    });
  });

  // Ensure all the imported globals get the first ids.
  const varResolver = new Utils.ScopedIdMap();
  importedVars.forEach(([poolName, variableName]) => {
    varResolver.get(poolName, variableName);
  });

  const functionImports = Object.entries(shims).map(([name, func]) => {
    return {
      args: new Array(func.length).fill(null).map(_ => VAL_TYPE.f64),
      // Shims implicitly always return a number
      returns: [VAL_TYPE.f64],
      name,
    };
  });

  const localFuncOrder: string[] = [];

  const moduleFuncs: {
    binary: number[];
    exportName: string;
    args: never[];
    returns: never[];
    localVariables: number[];
  }[] = [];

  Object.entries(funcs).forEach(([name, { pool, code }]) => {
    if (pools[pool] == null) {
      const poolsList = Object.keys(pools);
      if (poolsList.length === 0) {
        throw new Error(
          `The function "${name}" was declared as using a variable ` +
            `pool named "${pool}" but no pools were defined.`
        );
      }
      throw new Error(
        `The function "${name}" was declared as using a variable ` +
          `pool named "${pool}" which is not among the variable ` +
          `pools defined. The defined variable pools are: ` +
          `${Utils.formatList(poolsList)}.`
      );
    }
    const ast = preParsed ? code : parse(code);
    if (typeof ast === "string") {
      // TODO: Change the API so this can be enforced by types
      throw new Error(
        "Got passed unparsed code without setting the preParsed flag"
      );
    }
    if (ast.type !== "SCRIPT") {
      throw new Error("Invalid AST");
    }
    if (ast.body.length === 0) {
      return;
    }
    const localVariables: number[] = [];
    const context: CompilerContext = {
      resolveVar: name => {
        // The `reg00`-`reg99` variables are special in that they are shared between all pools.
        if (/^reg\d\d$/.test(name)) {
          return varResolver.get(null, name);
        }
        return varResolver.get(pool, name);
      },
      resolveLocal: type => {
        // TODO: We could provide a way for the emitter to release a local
        // variable so that we can reuse it, much in the same way a traditional
        // compiler does in register allocation.
        localVariables.push(type);
        return localVariables.length - 1;
      },
      resolveFunc: name => {
        // If this is a shim, return the shim index.
        const shimdex = functionImports.findIndex(func => func.name === name);
        if (shimdex !== -1) {
          const call = op.call(shimdex);
          if (name === "rand" && eelVersion === 1) {
            return [...call, op.f64_floor];
          }
          return call;
        }

        // If it's not a shim and it's not a defined function, return null.
        // The emitter will generate a nice error.
        if (localFuncMap[name] == null) {
          return null;
        }
        let index = localFuncOrder.indexOf(name);
        if (index === -1) {
          localFuncOrder.push(name);
          index = localFuncOrder.length - 1;
        }
        return op.call(index + functionImports.length);
      },
      rawSource: code,
    };
    const binary = emit(ast, context);

    moduleFuncs.push({
      binary,
      exportName: name,
      args: [],
      returns: [],
      localVariables,
    });
  });

  const localFuncs = localFuncOrder.map(name => {
    const func = localFuncMap[name];
    // This check is technically redundant since we check inside resolveLocalFunc
    // in the compiler context. It's here just to catch potential compiler bugs.
    if (func == null) {
      throw new Error(`Undefined local function "${name}"`);
    }
    return func;
  });

  // Given a function definition, return a hashable string representation of its signature.
  const getSignatureKey = (func: TypedFunction) => {
    return [...func.args, "|", ...func.returns].join("-");
  };

  // https://webassembly.github.io/spec/core/binary/modules.html#type-section
  const types: number[][] = [];
  const typeIndexByKey: Map<string, number> = new Map();
  [...functionImports, ...localFuncs, ...moduleFuncs].forEach(func => {
    const key = getSignatureKey(func);
    if (typeIndexByKey.has(key)) {
      return;
    }
    types.push([
      FUNCTION_TYPE,
      ...encodeFlatVector(func.args),
      ...encodeFlatVector(func.returns),
    ]);
    typeIndexByKey.set(key, types.length - 1);
  });

  function getTypeIndex(func: TypedFunction): number {
    const key = getSignatureKey(func);
    const typeIndex = typeIndexByKey.get(key);
    if (typeIndex == null) {
      throw new Error(`Failed to get a type index for key ${key}`);
    }
    return typeIndex;
  }

  // https://webassembly.github.io/spec/core/binary/modules.html#import-section
  const imports = [
    ...importedVars.map(([namespace, name]) => {
      return [
        ...encodeString(namespace),
        ...encodeString(name),
        ...[GLOBAL_TYPE, VAL_TYPE.f64, MUTABILITY.var],
      ];
    }),
    ...functionImports.map((func, i) => {
      const typeIndex = getTypeIndex(func);
      return [
        ...encodeString("shims"),
        ...encodeString(func.name),
        ...[TYPE_IDX, ...unsignedLEB128(typeIndex)],
      ];
    }),
  ];

  // https://webassembly.github.io/spec/core/binary/modules.html#function-section
  //
  // > Functions are referenced through function indices, starting with the smallest
  // > index not referencing a function import.
  const functions = [...localFuncs, ...moduleFuncs].map(func => {
    const typeIndex = getTypeIndex(func);
    return unsignedLEB128(typeIndex);
  });

  const memories = [
    // Only one memory
    [
      0x01, // Indicates that we are specifying two values (initial/max)
      ...unsignedLEB128(WASM_MEMORY_SIZE), // Initial size
      ...unsignedLEB128(WASM_MEMORY_SIZE), // Max size
    ],
  ];

  // https://webassembly.github.io/spec/core/binary/modules.html#global-section
  const globalCount = varResolver.size() - importedVars.length;
  const globals = Utils.times(globalCount, () => {
    return [
      VAL_TYPE.f64, // All eel values are float 64s
      MUTABILITY.var, // All globals are mutable
      ...op.f64_const(0), // Initialize the global to zero
      op.end, // All done
    ];
  });

  // https://webassembly.github.io/spec/core/binary/modules.html#binary-exportsec
  const xports = [...moduleFuncs].map((func, i) => {
    const funcIndex = i + functionImports.length + localFuncs.length;
    return [
      ...encodeString(func.exportName),
      EXPORT_TYPE.FUNC,
      ...unsignedLEB128(funcIndex),
    ];
  });

  /* Uncomment this to expose memory
  xports.push([
    ...encodeString("memory"),
    EXPORT_TYPE.MEMORY,
    ...unsignedLEB128(0),
  ]);
  */

  // https://webassembly.github.io/spec/core/binary/modules.html#code-section
  const codes = [...localFuncs, ...moduleFuncs].map(func => {
    // TODO: We could collapse consecutive types here, or even move to a two
    // pass approach where ids are resolved after the emitter is run.
    const localTypes = (func.localVariables ?? []).map(type => {
      return [...unsignedLEB128(1), type];
    });
    // It's a bit odd that every other section is an array of arrays and this
    // one is an array of vectors. The spec says this is so that when navigating
    // the binary functions can be skipped efficiently.
    return encodeFlatVector([
      ...encodeNestedVector(localTypes),
      ...func.binary,
      op.end,
    ]);
  });
  return new Uint8Array([
    // Magic module header
    ...MAGIC,
    // Version number
    ...WASM_VERSION,
    ...encodeSection(SECTION.TYPE, types),
    ...encodeSection(SECTION.IMPORT, imports),
    ...encodeSection(SECTION.FUNC, functions),
    ...encodeSection(SECTION.MEMORY, memories),
    ...encodeSection(SECTION.GLOBAL, globals),
    ...encodeSection(SECTION.EXPORT, xports),
    ...encodeSection(SECTION.CODE, codes),
  ]);
}
