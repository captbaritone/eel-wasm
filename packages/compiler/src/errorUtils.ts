import { SourceLocation } from "./types";

function repeat(n: number, char: string) {
  return new Array(n).fill(char).join("");
}

/* Build up a string showing a formatted source location in context with line numbers.
 *
 *   2 | z = sqr(y, 10);
 * > 3 | x = wat();
 *     |     ^^^
 *   4 | y = 100;
 */

export function printLoc(
  loc: SourceLocation,
  rawSource: string,
  contextLines = 1
) {
  const firstIndex = Math.max(loc.first_line - 1 - contextLines, 0);
  const lastIndex = loc.last_line + contextLines;

  const sourceLines = rawSource.split("\n").slice(firstIndex, lastIndex);

  const annotatedLines = sourceLines.map((line, i) => {
    const lineNumber = i + firstIndex + 1;
    const inRange = lineNumber >= loc.first_line && lineNumber <= loc.last_line;
    const gutter = inRange ? ">" : " ";
    return `${gutter} ${lineNumber} | ${line}`;
  });

  if (loc.first_line === loc.last_line) {
    const padding = repeat(loc.first_column, " ");
    const underline = repeat(loc.last_column - loc.first_column, "^");
    const insertIndex = loc.first_line - firstIndex;
    annotatedLines.splice(insertIndex, 0, `    | ${padding}${underline}`);
  }

  return annotatedLines.join("\n");
}

class CompilerError extends Error {
  loc: SourceLocation;
  sourceContext: string;

  constructor(message: string, loc: SourceLocation, rawSource: string) {
    super(message);
    // TODO: Create an error message that encourages users to open an issue at
    // https://github.com/captbaritone/eel-wasm/issues is they see this, and gives
    // Them an easy way to attach the right context.
    this.sourceContext = printLoc(loc, rawSource);
    this.loc = loc;
  }
}

class UserError extends CompilerError {}

export function createUserError(
  message: string,
  loc: SourceLocation,
  rawSource: string
) {
  return new UserError(message, loc, rawSource);
}

export function createCompilerError(
  message: string,
  loc: SourceLocation,
  rawSource: string
) {
  return new CompilerError(message, loc, rawSource);
}
