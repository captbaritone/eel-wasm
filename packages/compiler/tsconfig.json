{
  "compilerOptions": {
    "outDir": "./dist",
    "allowJs": true,
    "noImplicitAny": true,
    "skipLibCheck": true,
    "downlevelIteration": true,
    "sourceMap": true,
    "strictNullChecks": true
  },
  "include": ["./src/**/*"]
}
