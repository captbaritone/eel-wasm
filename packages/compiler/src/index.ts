import { parse } from "./parser";
import { compileModule } from "./compiler";
import shims from "./shims";
import { loadModule } from "./evaluator";

export { parse, compileModule, shims, loadModule };
