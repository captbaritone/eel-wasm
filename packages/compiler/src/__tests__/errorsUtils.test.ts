import fs from "fs";
import path from "path";
import { compileModule } from "../compiler";
import { printLoc } from "../errorUtils";

const DIR = path.join(__dirname, "../../fixtures/errors");

const testCases = fs.readdirSync(DIR);

testCases.forEach((filename: string) => {
  const testFunc = filename.startsWith("only.") ? test.only : test;
  const testName = filename.replace(/^only\./, "");
  testFunc(`${testName}`, () => {
    const eel = fs.readFileSync(path.join(DIR, filename), { encoding: "utf8" });
    let compilerError = null;
    try {
      compileModule({
        pools: { main: { globals: new Set(), functions: { run: eel } } },
      });
    } catch (e) {
      compilerError = e;
    }

    // TODO: Assert that error is a user or compiler error
    expect(compilerError).not.toBe(null);

    expect(compilerError.message).toMatchSnapshot();
    // We prepend a newline here so that the error starts on a newline rather
    // than a '"' in the snapshot and our underline alinment is easy to see.
    expect("\n" + compilerError.sourceContext).toMatchSnapshot();
  });
});

describe("printLoc", () => {
  test("One line", () => {
    const frame = printLoc(
      { first_line: 1, first_column: 6, last_line: 1, last_column: 11 },
      "Hello there this is a string"
    );
    expect("\n" + frame + "\n").toMatchInlineSnapshot(`
      "
      > 1 | Hello there this is a string
          |       ^^^^^
      "
    `);
  });
  test("One line within multiple lines", () => {
    const frame = printLoc(
      { first_line: 4, first_column: 6, last_line: 4, last_column: 11 },
      [
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
      ].join("\n")
    );
    expect("\n" + frame + "\n").toMatchInlineSnapshot(`
      "
        3 | Hello there this is a string
      > 4 | Hello there this is a string
          |       ^^^^^
        5 | Hello there this is a string
      "
    `);
  });
  test("One line within multiple lines + 2 lines of context", () => {
    const frame = printLoc(
      { first_line: 4, first_column: 6, last_line: 4, last_column: 11 },
      [
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
      ].join("\n"),
      2
    );
    expect("\n" + frame + "\n").toMatchInlineSnapshot(`
      "
        2 | Hello there this is a string
        3 | Hello there this is a string
      > 4 | Hello there this is a string
          |       ^^^^^
        5 | Hello there this is a string
        6 | Hello there this is a string
      "
    `);
  });
  test("Multiple lines", () => {
    const frame = printLoc(
      { first_line: 1, first_column: 6, last_line: 2, last_column: 11 },
      "Hello there this is a string\nHello there this is a string"
    );
    expect("\n" + frame + "\n").toMatchInlineSnapshot(`
      "
      > 1 | Hello there this is a string
      > 2 | Hello there this is a string
      "
    `);
  });

  test("Multiple lines within multiple lines", () => {
    const frame = printLoc(
      { first_line: 3, first_column: 6, last_line: 5, last_column: 11 },
      [
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
        "Hello there this is a string",
      ].join("\n")
    );
    expect("\n" + frame + "\n").toMatchInlineSnapshot(`
      "
        2 | Hello there this is a string
      > 3 | Hello there this is a string
      > 4 | Hello there this is a string
      > 5 | Hello there this is a string
        6 | Hello there this is a string
      "
    `);
  });
});
