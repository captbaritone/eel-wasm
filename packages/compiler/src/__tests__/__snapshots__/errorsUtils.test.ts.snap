// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assign_non_identifier.eel 1`] = `"Expected the first argument of \`assign()\` to be an identifier."`;

exports[`assign_non_identifier.eel 2`] = `
"
> 1 | assign(10 - 2, 10);
    |        ^^^^^^"
`;

exports[`assign_to_buffer_too_many_args.eel 1`] = `"Expected 1 argument when assinging to a buffer but got 2."`;

exports[`assign_to_buffer_too_many_args.eel 2`] = `
"
> 1 | megabuf(2, 3) = 10;
    |            ^"
`;

exports[`assign_to_non_buff_function_call.eel 1`] = `"The only function calls which may be assigned to are \`gmegabuf()\` and \`megabuf()\`."`;

exports[`assign_to_non_buff_function_call.eel 2`] = `
"
> 1 | abs(10) = 10;
    | ^^^"
`;

exports[`parse_error.eel 1`] = `"Parse Error: Expecting 'EOF', got ')'"`;

exports[`parse_error.eel 2`] = `
"
  2 | z = sqr(y, 10);
> 3 | x = assign);
    |     ^^^^^^
  4 | y = 100;
"
`;

exports[`too_few_arguments.eel 1`] = `"Too few arguments passed to \`pow()\`. Expected 2 but only got 1."`;

exports[`too_few_arguments.eel 2`] = `
"
> 1 | x = pow(1);
    |     ^^^^^^"
`;

exports[`too_many_arguments.eel 1`] = `"Too many arguments passed to \`pow()\`. Expected 2 but got 3."`;

exports[`too_many_arguments.eel 2`] = `
"
> 1 | x = pow(1, 2, 3);
    |               ^"
`;

exports[`undefined_function.eel 1`] = `"\\"noexist\\" is not defined."`;

exports[`undefined_function.eel 2`] = `
"
> 1 | x = noexist();
    |     ^^^^^^^"
`;

exports[`undefined_function_multiline.eel 1`] = `"\\"noexist\\" is not defined."`;

exports[`undefined_function_multiline.eel 2`] = `
"
  2 | z = sqr(y);
> 3 | x = noexist();
    |     ^^^^^^^
  4 | y = 100;
"
`;

exports[`undefined_function_with_comments.eel 1`] = `"\\"noexist\\" is not defined."`;

exports[`undefined_function_with_comments.eel 2`] = `
"
  3 | // Comment line three
> 4 | x = noexist();
    |     ^^^^^^^"
`;
